feat!: support running tests using VM context (#3203)
vitest-dev/vitest: docs/config/index.md

#### deps.external

- **Type:** `(string | RegExp)[]`
- **Default:** `[]`

> Externalize means that Vite will bypass the package to native Node. Externalized dependencies will not be applied Vite's transformers and resolvers, so they do not support HMR on reload. By default, all packages inside `node_modules` are externalized.

Reviewer: ```suggestion
Externalize means that Vite will bypass the package to the native Node. Externalized dependencies will not be applied to Vite's transformers and resolvers, so they do not support HMR on reload. By default, all packages inside `node_modules` are externalized.
```

Write test results to a file when the `--reporter=json`, `--reporter=html` or `--reporter=junit` option is also specified.
By providing an object instead of a string you can define individual outputs when using multiple reporters.

### experimentalVmThreads

- **Type:** `boolean`
- **CLI:** `--experimentalVmThreads`, `--experimental-vm-threads`
- **Version:** Since Vitest 0.34.0

Run tests using [VM context](https://nodejs.org/api/vm.html) (inside a sandboxed environment) in a worker pool.

This makes tests run faster, but VM module is unstable when running [ESM code](https://github.com/nodejs/node/issues/37648). Your tests will [leak memory](https://github.com/nodejs/node/issues/33439) - to battle that, consider manually editing [`experimentalVmWorkerMemoryLimit`](#experimentalvmworkermemorylimit) value.

::: warning
Running code in a sandbox has some advantages (faster tests), but also comes with a number of disadvantages.

> - Globals from native modules (`fs`, `path`, etc) are different from the globals in your test environment. It means that error thrown by any of the native modules will be different from the one you are using in your code:

Reviewer: @sheremet-va you're right with the second paragraph. It should be "the ones that were thrown in your code" for clarity.

> - Globals from native modules (`fs`, `path`, etc) are different from the globals in your test environment. It means that error thrown by any of the native modules will be different from the one you are using in your code:

Reviewer: "a distinct nature" also means "unique behavior" but it doesn't make any reference so it can confuse. Let's use "unrelated error instances." or "divergent error constructors."?

> - Globals from native modules (`fs`, `path`, etc) are different from the globals in your test environment. It means that error thrown by any of the native modules will be different from the one you are using in your code:

Author: Can't we use more technical term? This is not a poem. It should be understandable from reading it the first time - error from `fs.readFileSync` won't be instance of Error available in the test context

```ts
try {
  fs.writeFile('/doesnt exist')
}
catch (err) {

>   consoel.log(err instanceof Error) // false

Reviewer 2: I am pretty sure this is meant to be `console.log` :)

Reviewer: Oh yeah, Thanks

}
```

- Importing ES modules caches them indefinetly which introduces memory leak if you have a lot of contexts (or tests). There is no API in Node.js that clears that cache.
- Accessing globals [takes longer](https://github.com/nodejs/node/issues/31658) in a sandbox environment.

Please, be aware of these issues when using this option. Vitest team cannot fix any of the issues on our side.
:::

### experimentalVmWorkerMemoryLimit

- **Type:** `string | number`
- **CLI:** `--experimentalVmWorkerMemoryLimit`, `--experimental-vm-worker-memory-limit`
- **Default:** `1 / CPU Cores`
- **Version:** Since Vitest 0.34.0

> Specifies the memory limit for workers before they are recycled. This value heavily depends on your environment, so it's better to specify it manually instead of relying on default.

Reviewer: ```suggestion
Specifies the memory limit for workers before they are recycled. This value heavily depends on your environment, so it's better to specify it manually instead of relying on the default.
```

By providing an object instead of a string you can define individual outputs when using multiple reporters.

### experimentalVmThreads

- **Type:** `boolean`
- **CLI:** `--experimentalVmThreads`, `--experimental-vm-threads`
- **Version:** Since Vitest 0.34.0

Run tests using [VM context](https://nodejs.org/api/vm.html) (inside a sandboxed environment) in a worker pool.

This makes tests run faster, but the VM module is unstable when running [ESM code](https://github.com/nodejs/node/issues/37648). Your tests will [leak memory](https://github.com/nodejs/node/issues/33439) - to battle that, consider manually editing [`experimentalVmWorkerMemoryLimit`](#experimentalvmworkermemorylimit) value.

::: warning
Running code in a sandbox has some advantages (faster tests), but also comes with a number of disadvantages.

- The globals within native modules, such as (`fs`, `path`, etc), differ from the globals present in your test environment. As a result, any error thrown by these native modules will reference a different Error constructor compared to the one used in your code:

```ts
try {

>   fs.writeFile('/doesnt exist')

Author: ```suggestion
  fs.writeFileSync('/doesnt exist')
```

}
catch (err) {
  console.log(err instanceof Error) // false
}
```

> - Importing ES modules cache them indefinitely which introduces memory leaks if you have a lot of contexts (or tests). There is no API in Node.js that clears that cache.

Author: ```suggestion
- Importing ES modules caches them indefinitely which introduces memory leaks if you have a lot of contexts (test files). There is no API in Node.js that clears that cache.
```
